src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs

```diff
@@ -36,6 +40,96 @@ public static IServiceCollection AddMvc([NotNull] this IServiceCollection servic
+        /// in the <paramref name="services"/> and used for controller discovery.</param>
+        /// <param name="configuration">The application's <see cref="IConfiguration"/>.</param>
+        /// <returns>The <see cref="IServiceCollection"/>.</returns>
+        public static IServiceCollection WithControllersAsServices(
+            [NotNull] this IServiceCollection services,
+            [NotNull] IEnumerable<Type> controllerTypes,
+            IConfiguration configuration)
+        {
+            var controllerTypeInfos = new List<TypeInfo>();
+            foreach (var type in controllerTypes)
+            {
+                services.AddTransient(type);
+                controllerTypeInfos.Add(type.GetTypeInfo());
+            }
+
+            var describer = new ServiceDescriber(configuration);
+            services.Replace(describer.Transient<IControllerActivator, ServiceBasedControllerActivator>());
+            var controllerTypeProvider = new StaticControllerTypeProvider(controllerTypeInfos);
+            services.Replace(describer.Instance<IControllerTypeProvider>(controllerTypeProvider));
+
+            return services;
+        }
+
+        /// <summary>
+        /// Registers controller types from the specified <paramref name="assemblies"/> as services and as a source
+        /// for controller discovery.
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceCollection"/>.</param>
+        /// <param name="controllerAssemblies">Assemblies to scan.</param>
+        /// <returns>The <see cref="IServiceCollection"/>.</returns>
+        public static IServiceCollection WithControllersAsServices(
+            [NotNull] this IServiceCollection services,
+            [NotNull] IEnumerable<Assembly> controllerAssemblies)
+        {
+            return WithControllersAsServices(services,
+                                             controllerAssemblies,
+                                             loggerFactory: null,
+                                             configuration: null);
+        }
+
+        /// <summary>
+        /// Registers controller types from the specified <paramref name="assemblies"/> as services and as a source
+        /// for controller discovery.
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceCollection"/>.</param>
+        /// <param name="controllerAssemblies">Assemblies to scan.</param>
+        /// <param name="configuration">The application's <see cref="IConfiguration"/>.</param>
+        /// <param name="loggerFactory">The <see cref="ILoggerFactory"/>.</param>
+        /// <returns>The <see cref="IServiceCollection"/>.</returns>
+        public static IServiceCollection WithControllersAsServices(
+            [NotNull] this IServiceCollection services,
+            [NotNull] IEnumerable<Assembly> controllerAssemblies,
+            IConfiguration configuration,
+            ILoggerFactory loggerFactory)
```

Comment:
OK this is super-extra-double-broken. Need to come up with a better alternative to passing the logger in here.